{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "abf88bf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcfpy\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b316916d",
   "metadata": {},
   "source": [
    "# Quality filtering of mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3ac035e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/external.data/MeisterLab/Kalyan/EMS_sequencing_data/annotation/freebayes/EMS_1_CROSS_vs_PMW1074/EMS_1_CROSS_vs_PMW1074.freebayes.filtered.norm.sorted_snpEff.ann.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "work_dir=\"/mnt/external.data/MeisterLab/Kalyan/EMS_sequencing_data/\"\n",
    "out_dir=os.path.join(work_dir,\"QualityFilt\")\n",
    "\n",
    "vcf_dirs =[d for d in os.listdir(os.path.join(work_dir,\"annotation/freebayes\")) if os.path.isdir(os.path.join(work_dir,\"annotation/freebayes\",d))]\n",
    "for vcf_dir in vcf_dirs[0:1]:\n",
    "    vcf_path=os.path.join(work_dir+\"annotation/freebayes\",vcf_dir,vcf_dir+\".freebayes.filtered.norm.sorted_snpEff.ann.vcf.gz\")\n",
    "    print(vcf_path)\n",
    "    reader = vcfpy.Reader.from_path(vcf_path)\n",
    "    os.makedirs(os.path.join(out_dir,vcf_dir), exist_ok=True)\n",
    "    out_file=os.path.join(out_dir,vcf_dir,vcf_dir+\".freebayes.filtered.norm.sorted_snpEff.ann.vcf.gz\")\n",
    "    # with vcfpy.Writer.from_path(out_file, reader.header) as writer:\n",
    "    #     for record in reader:\n",
    "    #         ref = record.REF\n",
    "    #         alt = record.ALT[0].value\n",
    "            \n",
    "    #         if (ref == 'G' and alt == 'A') or (ref == 'C' and alt == 'T'):\n",
    "    #             writer.write_record(record)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87cfef37",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'Reader' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[4]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m reader\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m \u001b[43mreader\u001b[49m\u001b[43m[\u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[31mTypeError\u001b[39m: 'Reader' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "reader\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "522fa87f",
   "metadata": {},
   "source": [
    "# Filter only EMS mutations\n",
    "\n",
    "Used for plotting allele freq along chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a30aee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir=\"/mnt/external.data/MeisterLab/Kalyan/EMS_sequencing_data/\"\n",
    "out_dir=os.path.join(work_dir,\"EMSmutFilt\")\n",
    "\n",
    "\n",
    "vcf_dirs =[d for d in os.listdir(os.path.join(work_dir,\"annotation/freebayes\")) if os.path.isdir(os.path.join(work_dir,\"annotation/freebayes\",d))]\n",
    "for vcf_dir in vcf_dirs[0:3]:\n",
    "    vcf_path=os.path.join(work_dir+\"annotation/freebayes\",vcf_dir,vcf_dir+\".freebayes.filtered_snpEff.ann.vcf\")\n",
    "    print(vcf_path)\n",
    "    reader = vcfpy.Reader.from_path(vcf_path)\n",
    "    os.makedirs(os.path.join(out_dir,vcf_dir), exist_ok=True)\n",
    "    out_file=os.path.join(out_dir,vcf_dir,vcf_dir+\".freebayes.filtered_snpEff.ann.vcf\")\n",
    "    with vcfpy.Writer.from_path(out_file, reader.header) as writer:\n",
    "        for record in reader:\n",
    "            ref = record.REF\n",
    "            alt = record.ALT[0].value\n",
    "            \n",
    "            if (ref == 'G' and alt == 'A') or (ref == 'C' and alt == 'T'):\n",
    "                writer.write_record(record)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3df85064",
   "metadata": {},
   "source": [
    "# Create CSV with allele freq only for EMS mutations\n",
    "\n",
    "Used for plotting frequency along genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62398ce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir=\"/mnt/external.data/MeisterLab/Kalyan/EMS_sequencing_data/TvNmode/\"\n",
    "out_dir=os.path.join(work_dir,\"EMSmutFilt\")\n",
    "\n",
    "#  Extract VCF data into a table\n",
    "\n",
    "vcf_dirs =[d for d in os.listdir(os.path.join(work_dir,\"annotation/freebayes\")) if os.path.isdir(os.path.join(work_dir,\"annotation/freebayes\",d))]\n",
    "\n",
    "for vcf_dir in vcf_dirs[0:3]:\n",
    "    data = []\n",
    "    vcf_path = os.path.join(work_dir, \"annotation/freebayes\", vcf_dir, vcf_dir + \".freebayes.filtered_snpEff.ann.vcf\")\n",
    "    reader = vcfpy.Reader.from_path(vcf_path)\n",
    "    \n",
    "    for record in reader:\n",
    "        ref = record.REF\n",
    "        alt = record.ALT[0].value if record.ALT else \".\"\n",
    "        sample_names = reader.header.samples.names\n",
    "        \n",
    "        # Only process EMS mutations\n",
    "        if (ref == 'G' and alt == 'A') or (ref == 'C' and alt == 'T'):\n",
    "            # Extract AD field from first sample\n",
    "            ad = [call.data.get(\"AD\") or \".,.\" for call in record.calls]\n",
    "            \n",
    "            data.append({\n",
    "                'CHROM': record.CHROM,\n",
    "                'POS': record.POS,\n",
    "                'REF': ref,\n",
    "                'ALT': alt,\n",
    "                sample_names[0]+\"_refCount\": ad[0][0],\n",
    "                sample_names[0]+\"_altCount\": ad[0][1],  \n",
    "                sample_names[1]+\"_refCount\": ad[1][0],\n",
    "                sample_names[1]+\"_altCount\": ad[1][1]\n",
    "            })\n",
    "    df = pd.DataFrame(data)\n",
    "    print(f\"\\nTotal EMS mutations: {len(df)}\")\n",
    "    print(df.head(10))\n",
    "    df.to_csv(os.path.join(out_dir, vcf_dir+\"_EMS_mutation_counts.csv\"), index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d59e9bc",
   "metadata": {},
   "source": [
    "# Create CSV with allele freq for all SNPs with good MQ\n",
    "\n",
    "Used for plotting frequency along genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2e11003d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Total high quality snps: 1634\n",
      "  CHROM     POS REF ALT   MQM  MQMR  tm580_PMW1074_refCount  \\\n",
      "0     I  229065   C   T  60.0  60.0                      14   \n",
      "1     I  229117   T   C  60.0  60.0                      19   \n",
      "2     I  229124   C   T  60.0  60.0                      22   \n",
      "3     I  229174   A   G  60.0  60.0                      22   \n",
      "4     I  229204   A   T  60.0  60.0                      17   \n",
      "5     I  229214   G   A  60.0  60.0                      19   \n",
      "6     I  229296   G   A  60.0  60.0                      13   \n",
      "7     I  229382   G   C  60.0  60.0                      14   \n",
      "8     I  229430   A   G  60.0  60.0                      16   \n",
      "9     I  229442   A   C  60.0  60.0                      19   \n",
      "\n",
      "   tm580_PMW1074_altCount  \n",
      "0                      18  \n",
      "1                      15  \n",
      "2                      14  \n",
      "3                      19  \n",
      "4                      18  \n",
      "5                      22  \n",
      "6                      24  \n",
      "7                      27  \n",
      "8                      24  \n",
      "9                      24  \n",
      "\n",
      "Total high quality snps: 2147\n",
      "  CHROM     POS REF ALT      MQM  MQMR  tm580_EMS_1_CROSS_refCount  \\\n",
      "0     I  105306   G   A  60.0000  60.0                          11   \n",
      "1     I  229065   C   T  59.9231  60.0                          20   \n",
      "2     I  229117   T   C  60.0000  60.0                          22   \n",
      "3     I  229124   C   T  60.0000  60.0                          23   \n",
      "4     I  229174   A   G  60.0000  60.0                          21   \n",
      "5     I  229204   A   T  60.0000  60.0                          17   \n",
      "6     I  229214   G   A  60.0000  60.0                          19   \n",
      "7     I  229296   G   A  60.0000  60.0                          23   \n",
      "8     I  229382   G   C  60.0000  60.0                          28   \n",
      "9     I  229430   A   G  60.0000  60.0                          26   \n",
      "\n",
      "   tm580_EMS_1_CROSS_altCount  \n",
      "0                          17  \n",
      "1                          26  \n",
      "2                          25  \n",
      "3                          25  \n",
      "4                          33  \n",
      "5                          31  \n",
      "6                          32  \n",
      "7                          26  \n",
      "8                          33  \n",
      "9                          36  \n"
     ]
    }
   ],
   "source": [
    "work_dir=\"/mnt/external.data/MeisterLab/Kalyan/EMS_sequencing_data/\"\n",
    "out_dir=os.path.join(work_dir,\"MQsnpFilt\")\n",
    "os.makedirs(out_dir, exist_ok=True)\n",
    "\n",
    "#  Extract VCF data into a table\n",
    "\n",
    "vcf_dirs =[d for d in os.listdir(os.path.join(work_dir,\"annotation/freebayes\")) if os.path.isdir(os.path.join(work_dir,\"annotation/freebayes\",d))]\n",
    "\n",
    "for vcf_dir in vcf_dirs[3:5]:\n",
    "    data = []\n",
    "    vcf_path = os.path.join(work_dir, \"annotation/freebayes\", vcf_dir, vcf_dir + \".freebayes.filtered_snpEff.ann.vcf\")\n",
    "    reader = vcfpy.Reader.from_path(vcf_path)\n",
    "    \n",
    "    for record in reader:\n",
    "        ref = record.REF\n",
    "        alt = record.ALT[0].value if record.ALT else \".\"\n",
    "        sample_names = reader.header.samples.names\n",
    "        # Extract AD field from first sample\n",
    "        ad = [call.data.get(\"AD\") or \".,.\" for call in record.calls]\n",
    "        # extract mutation type\n",
    "        type = record.INFO[\"TYPE\"]\n",
    "\n",
    "        if (record.INFO[\"TYPE\"][0]==\"snp\"):\n",
    "            # extract mean quality scores for alt and ref alleles\n",
    "            MQM = record.INFO.get(\"MQM\", 0)\n",
    "            MQMR = record.INFO.get(\"MQMR\", 0)\n",
    "            if MQM[0] >=30 and MQMR >=30:\n",
    "                data.append({\n",
    "                    'CHROM': record.CHROM,\n",
    "                    'POS': record.POS,\n",
    "                    'REF': ref,\n",
    "                    'ALT': alt,\n",
    "                    'MQM': MQM[0],\n",
    "                    'MQMR': MQMR,\n",
    "                    sample_names[0]+\"_refCount\": ad[0][0],\n",
    "                    sample_names[0]+\"_altCount\": ad[0][1]\n",
    "                })\n",
    "\n",
    "    df = pd.DataFrame(data)\n",
    "    print(f\"\\nTotal high quality snps: {len(df)}\")\n",
    "    print(df.head(10))\n",
    "    df.to_csv(os.path.join(out_dir, vcf_dir+\"_high_quality_snp_counts.csv\"), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "33a05cfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2102"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "c1798532",
   "metadata": {},
   "outputs": [
    {
     "ename": "IncorrectVCFFormat",
     "evalue": "Missing line starting with \"#CHROM\"",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mIncorrectVCFFormat\u001b[39m                        Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[64]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m reader = \u001b[43mvcfpy\u001b[49m\u001b[43m.\u001b[49m\u001b[43mReader\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfrom_path\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvcf_path\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/EMSforest/lib/python3.11/site-packages/vcfpy/reader.py:106\u001b[39m, in \u001b[36mReader.from_path\u001b[39m\u001b[34m(cls, path, tabix_path, record_checks, parsed_samples)\u001b[39m\n\u001b[32m    104\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    105\u001b[39m     f = \u001b[38;5;28mopen\u001b[39m(path, \u001b[33m\"\u001b[39m\u001b[33mrt\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m106\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mfrom_stream\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    107\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstream\u001b[49m\u001b[43m=\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    108\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpath\u001b[49m\u001b[43m=\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    109\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtabix_path\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtabix_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    110\u001b[39m \u001b[43m    \u001b[49m\u001b[43mrecord_checks\u001b[49m\u001b[43m=\u001b[49m\u001b[43mrecord_checks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    111\u001b[39m \u001b[43m    \u001b[49m\u001b[43mparsed_samples\u001b[49m\u001b[43m=\u001b[49m\u001b[43mparsed_samples\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    112\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/EMSforest/lib/python3.11/site-packages/vcfpy/reader.py:66\u001b[39m, in \u001b[36mReader.from_stream\u001b[39m\u001b[34m(cls, stream, path, tabix_path, record_checks, parsed_samples)\u001b[39m\n\u001b[32m     64\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m tabix_path \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m path:  \u001b[38;5;66;03m# pragma: no cover\u001b[39;00m\n\u001b[32m     65\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mMust give path if tabix_path is given\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m66\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mReader\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m     67\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstream\u001b[49m\u001b[43m=\u001b[49m\u001b[43mstream\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     68\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpath\u001b[49m\u001b[43m=\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     69\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtabix_path\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtabix_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     70\u001b[39m \u001b[43m    \u001b[49m\u001b[43mrecord_checks\u001b[49m\u001b[43m=\u001b[49m\u001b[43mrecord_checks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     71\u001b[39m \u001b[43m    \u001b[49m\u001b[43mparsed_samples\u001b[49m\u001b[43m=\u001b[49m\u001b[43mparsed_samples\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     72\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/EMSforest/lib/python3.11/site-packages/vcfpy/reader.py:140\u001b[39m, in \u001b[36mReader.__init__\u001b[39m\u001b[34m(self, stream, path, tabix_path, record_checks, parsed_samples)\u001b[39m\n\u001b[32m    138\u001b[39m \u001b[38;5;28mself\u001b[39m.parser = parser.Parser(stream, \u001b[38;5;28mself\u001b[39m.path, \u001b[38;5;28mself\u001b[39m.record_checks)\n\u001b[32m    139\u001b[39m \u001b[38;5;66;03m#: the Header\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m140\u001b[39m \u001b[38;5;28mself\u001b[39m.header = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mparser\u001b[49m\u001b[43m.\u001b[49m\u001b[43mparse_header\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparsed_samples\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/EMSforest/lib/python3.11/site-packages/vcfpy/parser.py:761\u001b[39m, in \u001b[36mParser.parse_header\u001b[39m\u001b[34m(self, parsed_samples)\u001b[39m\n\u001b[32m    759\u001b[39m     \u001b[38;5;28mself\u001b[39m._read_next_line()\n\u001b[32m    760\u001b[39m \u001b[38;5;66;03m# parse sample info line\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m761\u001b[39m \u001b[38;5;28mself\u001b[39m.samples = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_handle_sample_line\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparsed_samples\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    762\u001b[39m \u001b[38;5;66;03m# construct Header object\u001b[39;00m\n\u001b[32m    763\u001b[39m \u001b[38;5;28mself\u001b[39m.header = header.Header(header_lines, \u001b[38;5;28mself\u001b[39m.samples)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/EMSforest/lib/python3.11/site-packages/vcfpy/parser.py:777\u001b[39m, in \u001b[36mParser._handle_sample_line\u001b[39m\u001b[34m(self, parsed_samples)\u001b[39m\n\u001b[32m    775\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\" \"Check and interpret the \"##CHROM\" line and return samples\"\"\"\u001b[39;00m\n\u001b[32m    776\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m._line \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m._line.startswith(\u001b[33m\"\u001b[39m\u001b[33m#CHROM\u001b[39m\u001b[33m\"\u001b[39m):  \u001b[38;5;66;03m# pragma: no cover\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m777\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m exceptions.IncorrectVCFFormat(\u001b[33m'\u001b[39m\u001b[33mMissing line starting with \u001b[39m\u001b[33m\"\u001b[39m\u001b[33m#CHROM\u001b[39m\u001b[33m\"\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m    778\u001b[39m \u001b[38;5;66;03m# check for space before INFO\u001b[39;00m\n\u001b[32m    779\u001b[39m line = \u001b[38;5;28mself\u001b[39m._line.rstrip()\n",
      "\u001b[31mIncorrectVCFFormat\u001b[39m: Missing line starting with \"#CHROM\""
     ]
    }
   ],
   "source": [
    "reader = vcfpy.Reader.from_path(vcf_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "EMSforest",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
